### PR TITLE
feat: introduce Message, Serializer, Deserializer and Binding interfaces

### PR DESCRIPTION
**Pull request by lance**
## Proposed Changes

This PR introduces the `Message`, `Serializer` and `Deserializer`, and `Binding` interfaces used to convert a `CloudEvent` into a `Message` that can be sent across a transport protocol. The first protocol implemented for this is HTTP, and some of the functionality formerly in `src/transport/http` has been simplified, reduced and/or moved to `/src/messages/http`.  

Test for V1 events are in place. Conformance tests have been modified to use these new interfaces vs. the HTTP `Receiver` class.

### Example usage
In an express.js application, simply create a `Message` using the incoming request headers and body. Then call `HTTP.toEvent(message)`.

```js
app.post("/", (req, res) => {
  const message = {
    headers: req.headers,
    body: req.body
  };

  const event = HTTP.toEvent(message);
});
```

Of course, this can be inlined as: `HTTP.toEvent({ headers: req.headers, body: req.body });`.

Sending a `Message` should also be straightforward. Convert a `CloudEvent` to a message, and send it using Axios or any other HTTP transport library that allows setting headers and body.

```js
const message = HTTP.binary(event);
axios.post({
  url: 'https://event.receiver.com',
  headers: message.headers,
  body: message.body
});
```

- Fixes: https://github.com/cloudevents/sdk-javascript/issues/322
- Fixes: https://github.com/cloudevents/sdk-javascript/issues/295
- Version: 3.2.0

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -18,14 +18,19 @@ export {
...
+  Serializer,
+  HTTP,
   // From transport
   Emitter,
   Receiver,
-  Mode,
   Protocol,
   TransportOptions,
-  Headers,
-  headersFor,
```
Comment (lholmquist) [inline src/index.ts:28]: If we are going to release this as 3.2.0, then i think we need to still export `headersFor` and not remove it until 4.0

Comment (lance) [reply src/index.ts:28]: Ahh - good point...

**Comment by lance** [comment]
@cloudevents/sdk-javascript-maintainers this should be ready now with both v1 and v03 of the spec being tested. PTAL

File: examples/express-ex/index.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
-/* eslint-disable no-console */
+/* eslint-disable */
 
 const express = require("express");
-const {Receiver} = require("cloudevents");
```
Comment (lholmquist) [inline examples/express-ex/index.js:4]: Should we wait until the 3.2.0 release comes out before updating this example?

Comment (lance) [reply examples/express-ex/index.js:4]: Hmm - I suppose. I actually didn't mean to commit that. I was just using it to test out the new interfaces. I'm good either way. I can revert this file.

File: src/transport/receiver.ts [hunk 3]
```diff
@@ -36,64 +14,13 @@ export class Receiver {
    * @param {Object|JSON} body The body of the HTTP request
    * @return {CloudEvent} A new {CloudEvent} instance
    */
-  static accept(
-    headers: Headers,
-    body: string | Record<string, unknown> | CloudEventV1 | CloudEventV03 | undefined | null,
-  ): CloudEvent {
+  accept(headers: Headers, body: string | Record<string, unknown> | undefined | null): CloudEvent {
```
Comment (lholmquist) [inline src/transport/receiver.ts:17]: Should we have a deprecation message printout when this is called?

Is it also possible in Typescript to deprecate a function using the jsdoc annotations?

**Comment by lance** [comment]
@lholmquist I can revert those tests if you think they are needed. My problem with them is that they were not really integration tests, as they were testing implementation details and not the public facing API. Coverage doesn't actually go down when these are removed!

File: src/transport/receiver.ts (hunk 3 above)
Comment (lance) [reply src/transport/receiver.ts:18]: Yes - adding `@deprecated` results in doc output that looks like this:

image

**Comment by lholmquist** [comment]
>  they were testing implementation details and not the public facing API.

Ok, then it is probably fine to remove them

File: src/transport/receiver.ts (hunk 3 above)
Comment (lholmquist) [reply src/transport/receiver.ts:18]: Nice,  probably want add the same to the emitter?

**Comment by lance** [comment]
@lholmquist I think this is probably ready to land, but I'm not sure if your comments are an implicit approval or not.

**Comment by lholmquist** [comment]
oh sorry,  i think i forgot to press the approve button

**Pull request merged**